crates/admin/src/cluster_controller/service/state.rs

```diff
@@ -315,59 +318,113 @@ where
+    for partition in partition_ids {
+        let log_id = LogId::from(partition);
+        let current_trim_point = bifrost_admin.get_trim_point(log_id).await?;
+        trim_points.insert(log_id, current_trim_point);
+    }
+
+    Ok(trim_points)
+}
+
+fn compute_new_safe_trim_points(
+    cluster_state: Arc<ClusterState>,
+    trim_points: HashMap<LogId, Lsn>,
+) -> BTreeMap<LogId, (Lsn, PartitionId)> {
+    let mut partition_statuses: BTreeMap<
+        PartitionId,
+        BTreeMap<GenerationalNodeId, &PartitionProcessorStatus>,
+    > = BTreeMap::new();
+    let mut safe_trim_points = BTreeMap::new();
+
+    for node_state in cluster_state.nodes.values() {
+        match node_state {
+            NodeState::Alive(AliveNode {
+                generational_node_id,
+                partitions,
+                ..
+            }) => {
+                for (partition_id, partition_processor_status) in partitions.iter() {
+                    partition_statuses
+                        .entry(*partition_id)
+                        .or_default()
+                        .insert(*generational_node_id, partition_processor_status);
                 }
-            } else {
-                warn!("Stop automatically trimming log '{log_id}' because not all nodes are running a partition processor applying this log.");
+            }
+            NodeState::Dead(_) | NodeState::Suspect(_) => {
+                // todo(pavel): until we implement trim-gap support (https://github.com/restatedev/restate/issues/2247),
+                //  pause trimming unless we know the status of all nodes dead nodes might come
+                //  back and will be stuck if their applied lsn < trim point.
+                warn!("Automatic log trimming paused until all nodes are reporting applied LSN status. This can lead to increased disk usage on log servers.");
+                return safe_trim_points;
             }
         }
+    }
 
-        Ok(())
+    for (partition_id, processor_status) in partition_statuses.into_iter() {
+        let log_id = LogId::from(partition_id);
+
+        // We trust that if a single node from the cluster reports a partition's archived LSN,
+        // that this is accessible to all other nodes that may need it.
+        let max_archived_lsn = processor_status
+            .values()
+            .map(|s| s.last_archived_log_lsn.unwrap_or(Lsn::INVALID))
+            .max()
+            .unwrap_or(Lsn::INVALID);
+
+        let min_applied_lsn = processor_status
+            .values()
+            .map(|s| s.last_applied_log_lsn.unwrap_or(Lsn::INVALID))
+            .min()
+            .unwrap_or(Lsn::INVALID);
+
+        let current_trim_point = trim_points.get(&log_id).unwrap_or(&Lsn::INVALID).clone();
+
+        if max_archived_lsn == Lsn::INVALID {
+            warn!("Not trimming log '{log_id}' because no node is reporting a valid archived LSN");
+        } else if min_applied_lsn < current_trim_point {
```

Comment:
If the implementation works correctly, shouldn't `min_applied_lsn` be always smaller than the current trim point?

Reply:
Indeed, there is a mistake here, which I only fixed once I was on the airplane and couldn't push 😅 It's corrected locally but I'll park this for now.